hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlValidator.java

```diff
@@ -17,53 +17,284 @@
+                                   SqlNode exp, SelectScope scope, boolean includeSystemVars) {
         if (isHiddenColumn(exp, scope)) {
             return;
         }
 
         super.addToSelectList(list, aliases, fieldList, exp, scope, includeSystemVars);
     }
 
+    @Override
+    public RelDataType deriveType(SqlValidatorScope scope, SqlNode expression) {
+        RelDataType derived = super.deriveType(scope, expression);
+        assert derived != null;
+
+        if (derived.getSqlTypeName() == CHAR) {
+            // normalize CHAR to VARCHAR
+            derived = HazelcastTypeFactory.INSTANCE.createSqlType(VARCHAR, derived.isNullable());
+            setValidatedNodeType(expression, derived);
+        }
+
+        switch (expression.getKind()) {
+            case LITERAL:
+                return deriveLiteralType(derived, expression);
+
+            case CAST:
+                return deriveCastType(derived, scope, expression);
+
+            default:
+                return derived;
+        }
+    }
+
+    @Override
+    public void validateLiteral(SqlLiteral literal) {
+        validateLiteral(literal, getValidatedNodeType(literal));
+    }
+
+    @Override
+    public void validateCall(SqlCall call, SqlValidatorScope scope) {
+        // Enforce type derivation for all calls before validation. Calcite may
+        // skip it if a call has a fixed type, for instance AND always has
+        // BOOLEAN type, so operands may end up having no validated type.
+        deriveType(scope, call);
+        super.validateCall(call, scope);
+    }
+
+    @Override
+    protected SqlNode performUnconditionalRewrites(SqlNode node, boolean underFrom) {
+        SqlNode rewritten = super.performUnconditionalRewrites(node, underFrom);
+
+        if (rewritten != null && rewritten.isA(SqlKind.TOP_LEVEL)) {
+            // rewrite operators to Hazelcast ones starting at every top node
+            rewritten.accept(HazelcastOperatorTableVisitor.INSTANCE);
+        }
+
+        return rewritten;
+    }
+
+    private RelDataType deriveLiteralType(RelDataType derived, SqlNode expression) {
+        RelDataType known = knownNodeTypes.get(expression);
+        if (derived == known) {
+            return derived;
+        }
+
+        SqlLiteral literal = (SqlLiteral) expression;
+
+        if (HazelcastIntegerType.supports(typeName(derived)) && literal.getValue() != null) {
+            // Assign narrowest type to non-null integer literals.
+
+            derived = HazelcastIntegerType.deriveLiteralType(literal);
+            setKnownAndValidatedNodeType(expression, derived);
+        } else if (typeName(derived) == DECIMAL) {
+            // Assign DOUBLE type to any standalone floating point literal: the
+            // exact type is inferred later from the context in which the literal
+            // appears.
+
+            derived = HazelcastTypeFactory.INSTANCE.createSqlType(DOUBLE);
```

Comment:
Could you please clarify the reason of this DECIMAL -> DOUBLE conversion? I tried to disable this branch and several tests failed, such as 
```
Expected :DOUBLE
Actual   :DECIMAL(38, 38)
```
But from this failure, it is not clear why is this a problem in the first place. What would happen if we leave DECIMAL here unchanged?

Reply:
The idea is that there should be no decimals if there are no explicit decimal operands involved in the expression (decimal columns or something explicitly casted to decimal). So any standalone floating point literal receives double type initially and later, if it's involved in some decimal calculation, its type is upgraded to decimal.

Reply:
But we may lose the precision in this case? What is the motivation of preferring the inexact type instead of the exact one? From the user perspective, he writes the exact number.

Reply:
With decimals we still can loose precision since they have finite precision (38 digits). The motivation is described above: no explicit decimals gives no implicit decimals. What is your idea? Switch to decimals for floating point computations by default?

Reply:
Sorry, I still cannot get the motivation. The user provides an exact number, we convert it to inexact. This seems confusing. Consider the following two queries. Will they produce the same result?
```
SELECT 1.1 FROM table
```
```
SELECT * FROM (SELECT 1.1. FROM table)
```

Reply:
Other databases use a separate notation for exact and inexact numbers [1] [2]:
```
1.1    // exact
1.1.E5 // inexact
```

[1] https://docs.microsoft.com/ru-ru/sql/t-sql/data-types/constants-transact-sql?view=sql-server-ver15#float-and-real-constants
[2] https://dev.mysql.com/doc/refman/5.7/en/number-literals.html

Reply:
Now I got it, looks like that would require some significant changes in literal handling, converters and tests. Let me see what I can do.

Reply:
fixed in 45df152bfe1b24bfa12e322bc2f181da323127f9